Add support for RTC - phase 1

fiam opened the pull request:
Some work towards fixing #102. Only works with UBLOX7 GPS protocol for now.

fiam committed:
Initial support for RTC

- Add software RTC support in drivers/rtc.c
- Add MSP_SET_RTC with code 246. Sets the RTC from a unix timestamp
(int32_t) and a nanoseconds number (uint32_t).
- Display the current time in response to the CLI status command.

fiam committed:
Make BB pick the current time from the RTC

If the real time is available, put it on the existing BB header
"Log start datetime".
Note that the format has been slightly modified to include decimal
part in the seconds.

fiam committed:
Add support for retrieving RTC from GPS

For now, only the fake GPS reports the time. Drivers will need
to add support for it.

fiam committed:
Support reading GPS time from UBLOX7 GPS

digitalentity commented on src/main/blackbox/blackbox.c:
```diff
@@ -1706,19 +1709,29 @@ void blackboxUpdate(timeUs_t currentTimeUs)
-static char startDateTime[20] = "0000-01-01T00:00:00";
-const char *blackboxGetStartDateTime(void)
+
+static char started_buf[FORMATTED_DATE_TIME_BUFSIZE];
```
Please avoid using a static buffer here - there is no need to keep the date/time string.
You can allocate buffer on stack and format data just before writing it to log in `blackboxWriteSysinfo`. This will save us 20 bytes of RAM.

fiam replied:
@digitalentity I know, it was just because of the macro which calls into `blackboxGetStartDateTime()` and because this function was also exposed in the header. I didn't want to touch the macro nor the public function without some feedback first. Should I just remove `blackboxGetStartDateTime()` from the header?

digitalentity replied:
Yes, remove both `blackboxGetStartDateTime` and `blackboxSetStartDateTime`. Use a local buffer and RTC API to get the date/time in `blackboxWriteSysinfo`

digitalentity commented on src/main/drivers/rtc.c:
```diff
@@ -0,0 +1,136 @@
+
+bool rtc_set(timestamp_t ts)
+{
+    rtcMillisSet = millis();
```
Maybe we can avoid keeping an `rtcMillisSet` variable by calculating a `ts - millis()` - a timestamp for the moment where millis is zero.

fiam replied:
That's definitely a good idea.

digitalentity commented on src/main/drivers/rtc.c:
```diff
@@ -0,0 +1,136 @@
+// as an offset to count from.
+static timestamp_t rtcTsSet = {0,};
+
+timestamp_t timestamp_unix(int32_t unix, uint32_t nanos)
```
While returning a struct is permitted by C I don't think it's a good practice. Copying structs is a) slower and b) usually yields larger code.

martinbudden replied:
I concur. Returning structs is not good practice.

digitalentity commented on src/main/drivers/rtc.h:
```diff
@@ -0,0 +1,44 @@
+// functions below.
+typedef struct {
+    int32_t unix;
+    uint32_t nanos;
```
Why do we need nanos here? Why not keep only UNIX epoch time?

digitalentity commented on src/main/drivers/rtc.h:
```diff
@@ -0,0 +1,44 @@
+    uint8_t minutes;
+    // 0-59
+    uint8_t seconds;
+    uint32_t nanos;
```
Why do we need nanos for real-time clock?

digitalentity commented on src/main/drivers/rtc.c:
```diff
@@ -0,0 +1,136 @@
+    dt->hours = unix%24;
+    unix /= 24;
+
+    unsigned int years = unix/(365*4+1)*4;
```
Please keep the same style for all expressions. Here we usually add spaces: `unix / (365 * 4 + 1) * 4;`

digitalentity commented on src/main/blackbox/blackbox.c:
```diff
@@ -1706,19 +1709,29 @@ void blackboxUpdate(timeUs_t currentTimeUs)
+        dt.seconds = 0;
+        dt.nanos = 0;
+    }
+    date_time_format(started_buf, &dt);
```
If `rtc_get_dt` will return false `date_time_format` will format a random date/time - `dt` struct is allocated on stack and not zeroed.

fiam replied:
If `rtc_get_dt()` returns false, the code inside the `if` sets the year to zero, month to 1, etc... This doesn't format a random value, this makes the code return 0000-01-01... when there's no time available, just like it did before. I think this should be kept as is.

digitalentity replied:
Oops, I obviously overlooked that.

digitalentity replied:
On the other hand it may be desired to move that reature to RTC so `rtc_get_dt` will always provide valid data - return `false` will merely indicate that there's no time data available.

digitalentity commented on src/main/fc/cli.c:
```diff
@@ -2320,7 +2321,17 @@ static void cliStatus(char *cmdline)
+    if (rtc_get_dt(&dt)) {
+        char buf[FORMATTED_DATE_TIME_BUFSIZE];
+        date_time_format(buf, &dt);
+        cliPrint(buf);
```
Please replace this with `cliPrintLinef` - this way you won't need a separate call to `cliPrintLinefeed`

digitalentity commented on src/main/fc/fc_msp.c:
```diff
@@ -2113,6 +2114,14 @@ static mspResult_e mspFcProcessInCommand(uint8_t cmdMSP, sbuf_t *src)
         break;
 #endif
 
+    case MSP_SET_RTC:
```
In case of MSP it's allowed to set time multiple times, right?

fiam replied:
That's right, the current API always allows setting the time via MSP by design. If an external device sends a request to adjust the time, we just allow it.

digitalentity commented on src/main/io/gps.c:
```diff
@@ -146,6 +146,13 @@ void gpsSetState(gpsState_e state)
+static void gpsUpdateTime(void)
+{
+    if (!rtc_has_time() && gpsSol.flags.validTime) {
+        rtc_set_dt(&gpsSol.time);
```
I think GPS should call `rtc_set_dt` regardless and let RTC code decide wether it wants to update the time or not.

fiam replied:
Then we need to change the API to set the time, since we'd have calls which ask to always set the time (e.g. via MSP) and calls that are reporting the time they found from external sources and passing it to the RTC to decide what it wants to do with them (e.g. via GPS).

digitalentity replied:
The API should be the same. Maybe `rtcUpdateDateTime` - and RTC system will make a decision if the time should be updated.

digitalentity commented on src/main/drivers/rtc.c:
```diff
@@ -0,0 +1,136 @@
+        dt->hours, dt->minutes, dt->seconds, dt->nanos);
+}
+
+bool rtc_has_time()
```
I believe this API is redundant. All get function return `bool` to indicate success.

fiam replied:
This allows a caller to know if the RTC has already a time value before calling `rtc_set()`. We can change `rtc_get()` to allow passing a NULL pointer to implement the same functionality, but the intent seems more clear with a function which just returns whether there's already a valid time set.

digitalentity replied:
What's the use case for `rtc_has_time`? Functions using `rtc_get_dt` will need something valid regardless. Functions calling `rtc_set_dt` should do it regardless - decision about changing the clock should be made entirely within RTC system.

digitalentity commented on src/main/msp/msp_protocol.h:
```diff
@@ -319,6 +319,7 @@
 #define MSP_SERVO_MIX_RULES      241    //out message         Returns servo mixer configuration
 #define MSP_SET_SERVO_MIX_RULE   242    //in message          Sets servo mixer configuration
 #define MSP_SET_4WAY_IF          245    //in message          Sets 4way interface
+#define MSP_SET_RTC              246    //in message          Sets the RTC clock - TODO: Make sure this code is not used by other forks
```
Let's not touch MSPv1 messages. Please move RTC MSP API to `msp_protocol_v2_inav.h` header and change the ID according to https://github.com/iNavFlight/inav/wiki/MSP-V2#msp-v2-message-catalogue (change to range 0x1000 - 0x1FFF)

martinbudden replied:
I'm not sure about this. `MSP_SET_RTC` seems a valid v1 message, since it could be used by the Taranis to set the time. As such the code could be shared with betaflight. Currently neither Taranis nor Betaflight understand MSPv2.

digitalentity replied:
Good point. Let's keep it MSPv1 then for compatibility.

fiam replied:
MSPv1 it is then. Is 246 OK to use?

digitalentity replied:
Any unused value should do. Value of 246 seems to be unused across INAV, BF and CF

martinbudden commented on src/main/drivers/rtc.c:
```diff
@@ -0,0 +1,136 @@
+#include "common/printf.h"
```
Most of the functions in `rtc.c` seem to do with formatting time and time conversion. Such functions don't belong in a device driver, they should be in `common/time.c`

digitalentity replied:
I second this. Also definitions of `timestamp_t`, `date_time_t` and RTC handling functions should probably go to `common/time.h`

fiam replied:
@martinbudden A left over from the initial version, which was using the hardware RTC. I'll move everything somewhere where it makes more sense.

martinbudden commented on src/main/drivers/rtc.h:
```diff
@@ -0,0 +1,44 @@
+// functions below.
+typedef struct {
+    int32_t unix;
+    uint32_t nanos;
```
Note, that nanoseconds won't fit in a `uint32_t`.

Note also the iNav works in an extremely constrained environment - we have tight constraints on ROM size, RAM size and processor cycles. The easiest way to save these things is not to add unrequired functionality.

There is no current use case for nanosecond time. Especially for the RTC. Internal events are times using the system clock and will continue to be done so.

All nanosecond times should be removed. If we find we need them later we can add them - it should be straightforward to do this without breaking anything.

martinbudden replied:
`unix` is a bit of a confusing variable name. `unixTime` is unambiguous.

digitalentity replied:
I'd remove the `timestamp_t` struct altogether and do a `typedef unit32_t rtcTimeStamp_t;`

fiam replied:
@martinbudden nanoseconds fit in 30 bits unsigned, 31 with sign. log2(1e+9) is 29 something. The reason I went with nanoseconds is because using milliseconds will require 16 bits for them, so nanos only needs 16 more bits. Since the GPS reports nanosecond, 16 bits seem like small enough to not make a difference in size/ram usage while making the API future proof.

digitalentity replied:
YAGNI = You Ain't Going to Need It. 
What's the use case for nanosecond RTC? For measuring accurate intervals we use `micros()` or system ticks. For date/time - 1 second resolution is more than enough. Since there is no practical use for nanos in forseable future - it's safe to assume we don't need them implemented at all.

martinbudden commented on src/main/drivers/rtc.h:
```diff
@@ -0,0 +1,44 @@
+// buf must be at least FORMATTED_DATE_TIME_BUFSIZE
+void date_time_format(char *buf, date_time_t *dt);
+
+bool rtc_has_time();
```
Please use camelCase, ie `rtcHasTime`.

fiam committed:
Implement changes suggested during review

- Move implementation to common/time.
- Use camelCase for function names.
- Store just one value for the RTC, the real time when system
was started.
- Store real time as the number of milliseconds from 1970-01-01.
- Remove BB functions for manipulating time, query the RTC
directly when BB is started.

fiam committed:
Add license header to time.c
